circuit/src/ecc/mod.rs

```
@@ -2,628 +2,567 @@
-			config.integer.clone(),
+impl<W: FieldExt, N: FieldExt, const NUM_LIMBS: usize, const NUM_BITS: usize, P> Chipset<N>
+	for EccAddChipset<W, N, NUM_LIMBS, NUM_BITS, P>
+where
+	P: RnsParams<W, N, NUM_LIMBS, NUM_BITS>,
+{
+	type Config = EccAddConfig;
+	type Output = AssignedPoint<W, N, NUM_LIMBS, NUM_BITS, P>;
+
+	/// Synthesize the circuit.
+	fn synthesize(
+		self, common: &CommonConfig, config: &Self::Config, mut layouter: impl Layouter<N>,
+	) -> Result<Self::Output, Error> {
+		// Reduce p_x
+		let p_x = IntegerReduceChip::new(&self.p.x);
+		let p_x_reduced = p_x.synthesize(
+			&common,
+			&config.integer_reduce_selector,
 			layouter.namespace(|| "reduce_p_x"),
 		)?;
-		let p_y = IntegerChip::reduce(
-			p_y,
-			p_y_rw,
-			config.integer.clone(),
+
+		// Reduce p_y
+		let p_y = IntegerReduceChip::new(&self.p.y);
+		let p_y_reduced = p_y.synthesize(
+			&common,
+			&config.integer_reduce_selector,
 			layouter.namespace(|| "reduce_p_y"),
 		)?;
-		let q_x = IntegerChip::reduce(
-			q_x,
-			q_x_rw,
-			config.integer.clone(),
+
+		// Reduce q_x
+		let q_x = IntegerReduceChip::new(&self.q.x);
+		let q_x_reduced = q_x.synthesize(
+			&common,
+			&config.integer_reduce_selector,
 			layouter.namespace(|| "reduce_q_x"),
 		)?;
-		let q_y = IntegerChip::reduce(
-			q_y,
-			q_y_rw,
-			config.integer.clone(),
+
+		// Reduce q_y
+		let q_y = IntegerReduceChip::new(&self.q.y);
+		let q_y_reduced = q_y.synthesize(
+			&common,
+			&config.integer_reduce_selector,
 			layouter.namespace(|| "reduce_q_y"),
 		)?;
 
-		let (x, y) = Self::add_unreduced(
-			p_x,
-			p_y,
-			q_x,
-			q_y,
-			reduction_witnesses,
-			config,
-			layouter.namespace(|| "reduce_add"),
-		)?;
-		Ok((x, y))
+		// numerator = other.y.sub(&self.y);
+		let numerator_chip = IntegerSubChip::new(&q_y_reduced, &p_y_reduced);
+		let numerator = numerator_chip
+			.synthesize(
+				&common,
+				&config.integer_sub_selector,
+				layouter.namespace(|| "numerator"),
+			)
+			.unwrap();
```

We should propagate the error down the stack using `?`.
Apply to similar cases